### ISSUE TITLE
MSYS2: Support for MinGW64 Python should be added

### ISSUE THREAD

**Issue by overetou** [Nuitka/Nuitka#872]
Hello,

The problem was evoked in other issues (on nuitka and msys2) but the solution was never clearly given from what I have found (references in https://github.com/msys2/MINGW-packages/issues/4868 and #510 ->at some point zegervdv might have met the same problem)

I have tried the develop and the stable versions with the same results.

```
$ python3 -m nuitka --version
0.6.10rc5
Python: 3.8.6 (default, Oct  1 2020, 13:01:33)  [GCC 10.2.0 64 bit (AMD64)]
Executable: D:/Programmes/msys2/mingw64/bin/python3.exe
OS: Windows
Arch: x86_64
```

I did try to install the msys2/mingw package of nuitka but it complained that gcc.10 was < gcc.4.4 in version and refused to go further. Using an installation from pip (the one that returned its version above) seems to get me further, as the result of the compile command is now:

```
$ python3 -m nuitka --mingw64 test.py
Nuitka-Scons:INFO: Backend C compiler: gcc (gcc).
Error, no 'Python.h' development headers can be found at 'D:/Programmes/msys2/mingw64/include', dependency not satisfied!
```

test.py is a simple hello world.

I have removed pure msys2 python with pacman so that only mingw-w64-x86_64-python package remains, as suggested somewhere, but to no avail.

I cannot use mingw without msys2 or msys2 without mingw because my application uses pygobject, and this framework is only supported that way on windows to my knowledge.

```
$ python-config --ldflags --cflags
-LD:/Programmes/msys2/mingw64/lib [...]  -ID:/Programmes/msys2/mingw64/include/python3.8 [...]
```
There is indeed a Python.h file in Programmes/msys2/mingw64/include/python3.8/, but when I copy it inside D:/Programmes/msys2/mingw64/include/, nuitka then reports not finding "python38.lib" and that file I cannot find anywhere on my hard drives.

My final goal is here is to distribute my application on windows and nuitka seems perfect for the task (it is succesful on linux), so if there is a way to make it work I would be glad to know about it or to lend a hand to find it.

Thanks a lot,

Oscar

**Comment by kayhayen** [comment]
I believe gcc 10 is fine for stable Nuitka for a long time already, this wasn't the current stable 0.6.9 series, was it?

**Comment by kayhayen** [comment]
At some point, MSYS2 was working, but it keeps on changing and breaking a lot, so I had given up on it myself. It is a nice platform though, and potentially the best to use if you are foreign to Windows, and want to use Python on it. I think I uninstalled it, when I noticed that after an update, the ccache binary started hanging indefinitely. 

But since I kind of like it, and being ill don't get anything done anyway, I will retry myself. But claiming it as supported is probably not going to happen, it's too changing for that.

**Comment by kayhayen** [comment]
What's strange, on a fresh install, I am getting with `pacman -S python` and `pacman -S gcc` a `gcc-9.3.0-1` package and not a 10 version at all. Did you attempt to use some mingw package in pacman? I never used anything but gcc if I recall correctly.

Also, the `--mingw64` option should not be used, because well, you are not using it. It is a POSIX platform, and the gcc is not all needing to do any of those tricks needed there. It should pick up the `gcc` from the command line via `PATH` just fine, which it does for me. 

It even noticed the lack of `ccache.exe` and downloads it for use locally from the official project.

So I guess, this just works. And also, there is no `--mingw` option when running on MSYS2, so even more confusing to me, what you are doing there.

**Comment by kayhayen** [comment]
Yup, the option is guarded by `if os.name == "nt"`  which that is `posix` on MSYS2, so something wrong is doing on.

I guess, I could add instructions to the README of Nuitka, with a caveat. Also `python-devel` is needed, or else a header file for `Python.h` will be missing.

**Comment by overetou** [comment]
No problem Kay, thanks for the answer.

_this wasn't the current stable 0.6.9 series, was it?_

No indeed, they are distributing the 0.6.7-1 version. https://packages.msys2.org/package/mingw-w64-x86_64-python-nuitka?repo=mingw64

_Did you attempt to use some mingw package in pacman?_
Exactly. That's the sine qua non condition to use pygobject on windows, the official python bindings provider for GTK and some other projects. https://pygobject.readthedocs.io/en/latest/
Quite a convoluted setup, I concur.

_I am getting with pacman -S python and pacman -S gcc a gcc-9.3.0-1 package and not a 10 version at all_
You have installed this package: https://packages.msys2.org/package/gcc?repo=msys&variant=x86_64
and I am using this one: https://packages.msys2.org/package/mingw-w64-x86_64-gcc?repo=mingw64

_there is no --mingw option when running on MSYS2_
I installed and I run Nuitka on this package: https://packages.msys2.org/package/mingw-w64-x86_64-python?repo=mingw64
I must admit that I don't know the actual difference between msys2 python and msys2 + mingw python.
I used --mingw64 because otherwise no compiler is detected. Just happened to find the option on the user manual and it seemed to get me further. (Though my configuration is obviously different from what is described in the guide, it makes a bit of sense, since my setup in msys2 is exclusively made of mingw packages. But it's still a long shot.)

_python-devel is needed_
I guess that's where my problem lies. msys2 packages do provide a python-devel, but not with the mingw-ect prefix.

Nuitka is cool non the less, I will keep using it on linux.

Best,

Oscar

**Comment by kayhayen** [comment]
I see, I wasn't aware that Nuitka was packaged for msys2 at all. It's behind though, the current release is 0.6.9.4 and that is a lot of improvements missing, where gcc 10 is only one of many. 

The way to try your package would be to set `CC=/mingw64/bin/gcc` but that's only going to reveal the lack of this gcypt, as I also just tested, but that's probably all it takes. But I take, there is also a special Python, and that of course should be called, as cross compilation is not what should be done.

I wonder though, how they build there packages without a `Python.h` being present, there must be one hidden somewhere.

**Comment by kayhayen** [comment]
This other Python is not a POSIX one, but has `os.name == "nt"`, yet with non-standard file system layout, meh.

**Comment by kayhayen** [comment]
Ok, with a few tweaks, I got this to run just now, will be part of the next release, but I am waiting for a hotfix/pre-release to be made before pushing this to factory.

Basically, the `Python.h` to be used, is not where it's for CPython or Anaconda, more like a POSIX file structure, checking for both is no biggie though. And for the `Python.LIB`, there is a `libpython3.8.dll.a` that is usable, and it's like an uninstalled Python, because it won't find the `libpython3.8.dll` unless put alongside.

So that completes the puzzle. Unfortunately this is not really a static Python linkage, which I am always looking for.

**Comment by kayhayen** [comment]
Standalone mode works too, so that is nice. I assume module mode will need some work though, but nothing's really too hard about these things.

**Comment by overetou** [comment]
Great! I will make sure to try it when it releases.

**Comment by kayhayen** [comment]
It's on the factory branch. I think the only weakness is that you have to point it with CC=`` to use the matching `gcc.exe` since the one in the `PATH` will not work. Could automate that too.

**Comment by overetou** [comment]
Well done, it works on my machine!

I could compile a hello world on factory without problems, but not like you said, from what I understood :
`CC=/mingw64/bin/gcc.exe` does not help, Nuitka is still unable to find it without option (maybe I am doing something wrong).
However, with `--mingw64` it compiles and executes well.
My new limit are libraries. For instance, compiling with pygit2 or pygobject produces an executable without problem with this command: ` python -m nuitka --standalone --windows-dependency-tool=pefile test.py` but on execution, I get something like that:

> Traceback (most recent call last):
  File "D:\téléchargement\garden-of-mists\test.dist\test.py", line 1, in <module>
  File "D:\téléchargement\garden-of-mists\test.dist\pygit2/__init__.py", line 30, in <module pygit2>
ModuleNotFoundError: No module named 'pygit2._pygit2'

for this program, wich runs fine with the interpreter:

> import pygit2
> 
> print("hey!")

Same thing with my application:

> $ ./gardenofmists.exe
> Traceback (most recent call last):
>   File "D:\téléchargement\garden-of-mists\gardenofmists.dist\gardenofmists.py", line 2, in <module>
>   File "D:\téléchargement\garden-of-mists\gardenofmists.dist\custom_window.py", line 1, in <module custom_window>
>   File "D:\téléchargement\garden-of-mists\gardenofmists.dist\common.py", line 1, in <module common>
>   File "D:\téléchargement\garden-of-mists\gardenofmists.dist\gi/__init__.py", line 40, in <module gi>
> ImportError: cannot import name '_gi' from 'gi' (D:\téléchargement\garden-of-mists\gardenofmists.dist\gi/__init__.py)

Note: It seems that pygit2 has a problem with ssl certificates in a frozen environement (that's the reason why I discovered Nuitka, as an alternative to pyinstaller and co) and it may appear with nuitka too, but this problem likely comes from their side.

**Comment by kayhayen** [comment]
Make a new issue out of these, they seem typical stuff that might be dealt with independently. I am confused about the --mingw need and it working. Maybe the gcc.exe from there is in your path, because it's where the python binary is too. But I of course have 2 pythons and 2 gccs, that won't easily match inside my MSYS2 now, but would have to. If you only have one of them, it's going to be easier maybe.

Regarding --mingw, that's automatic if the `CC` is specified, and it causes it to search in PATH. But one is going to be broken by installation of the other, and I assume the POSIX python will be first in PATH, and so ought to the POSIX gcc. It's a strange thing anyway, to have 2 distinct things packaged in one, MSYS2 is bizarre with this.

**Comment by MaD70** [comment]
@kayhayen wrote:
> Unfortunately this is not really a static Python linkage, which I am always looking for.

If I understand you correctly, you are searching for something like [Python Standalone Builds](https://github.com/indygreg/python-build-standalone). Quoting from [Distributing Standalone Python Applications](https://gregoryszorc.com/blog/2018/12/18/distributing-standalone-python-applications/) blog post:
> I've created the python-build-standalone project for automating the process of building Python interpreters suitable for use with standalone, distributable Python applications. The project produces (and has available for download) binary artifacts including a pre-compiled Python interpreter and object files used for compiling that interpreter. **The Python interpreter is compiled with [PGO](https://en.wikipedia.org/wiki/Profile-guided_optimization)/[LTO](https://en.wikipedia.org/wiki/Interprocedural_optimization#WPO_and_LTO)** using a modern Clang, helping to ensure that Python code runs as fast as it can. All of Python's dependencies are compiled from source with the modern toolchain and **everything is aggressively statically linked to avoid external dependencies**. The toolchain and pre-built distribution are available for downstream consumers to compile Python extensions with/against.

I myself have no experience with it, I am not even a Python programmer: given the growing number of interesting software written in it (both applications and libraries), I am investigating the field of easy deployment (the less dependencies to install, the better) and good performance of Python software if the need should arise.

In this blog post he mentions Nuitka also (note that it was written in December 2018):
> Nuitka _converts_ Python source to C code then compiles and links that C code against libpython. You can perform a static link and compile everything down to a single executable. If you do the compiling properly, that executable should just work on pretty much every target machine. That's pretty cool and is exactly the kind of solution application distributors are looking for: you can't get much simpler than a self-contained executable! While I'd love to vouch for Nuitka and recommend using it, I haven't used it so can't. And I'll be honest, the prospect of compiling Python source to C code kind of terrifies me. That effectively makes Nuitka a new Python implementation and I'm not sure I can (yet) place the level of trust in Nuitka that I have for e.g. CPython and PyPy.

### Side note
I have a couple of questions/comments but I'm not sure it's appropriate to open github issues for them and I haven't found a mailing-list (in particular, point 2 is related to your request for a statically-linkable Python interpreter). I write them down here and you suggest me where I should put them (be polite! 😁):

1. in [Developer Manual](http://nuitka.net/doc/developer-manual.html#prefer-list-contractions-over-built-ins) it's used the term "_contraction_" which puzzled me a bit. In [a later section](http://nuitka.net/doc/developer-manual.html#list-contractions) there are code examples from which I understand that it's local jargon for what is universally known as [comprehension](https://en.wikipedia.org/wiki/List_comprehension) (even Python documentation uses it profusely, [here for example](https://docs.python.org/3/reference/expressions.html#parenthesized-forms)). Why this weirdness? Is it a deliberate choice or is it a translation problem (I'm not a native English speaker myself, so I'm sympathetic)?
2. I see that you're investing time in doing a lot of optimizations before generating C code and, knowing how much more capable are contemporary C compilers to this regard with respect to the past<sup>a</sup>, I'm curious to know what performance problems have you encountered with a more simple and straightforward translation approach.
I'm asking because exploring the assembly generated by GCC or Clang (sometimes already at `-O1` optimization level) I see that (a growing number of) functions called with constant parameters are computed at compile-time<sup>b</sup> and so reduced to constants and then, with LTO, propagated even across compilation units. Those functions with only some constant parameters are potentially simplified and in-lined; also, with LTO, functions/variables never used are never included in the executable (the standard link process doesn't have this level of granularity).
If one could compile and link statically, with LTO, both the Python interpreter and his/her own Python sources transcompiled to C, I think one can get some interesting performances (both in terms of execution time and in size of the executable). Reached this milestone you should be able to tweak Nuitka on two fronts: simplify it where C compilers are able to optimize with minimal (annotations) or no help and concentrate your effort on those Python features that don't translate straightforwardly to efficient C code.

Ok, I let myself got carried away a bit: I apologize if my observations in point 2 are trivial (or misguided) and you have already tried them or they present technical obstacles, that I neglected because I know nothing about the Python interpreter, which are difficult to overcome. The thing is, I just discovered this project and I see a lot of potential and a lot of value in it, so I hope it will be completely successful even if I haven't an immediate need for it.

**P.S.:** MSYS2 support is a good thing: in my experience is easy to install and upgrade, a lot of *nix tools make build scripts mostly reusable across platforms, and GCC (I think Clang also) is updated shortly after the upstream project update.
___
a) from exploiting program annotations like `const`, `inline`, `static`, `__attribute__ ((pure))`, and so on…, to compiler flags for [whole-program optimization or link-time optimization](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#index-fweb) — which optimizes across compilation units. Probably when you started this project common available C compilers where less sophisticated;
b) in C++ with `constexpr` and templates a programmer has better control of this (compile-time execution of code), even if templates are syntactically horrid. To see the power of `constexpr` at work, see this project: [constexpr-8cc: Compile-time C Compiler](https://github.com/keiichiw/constexpr-8cc).

**Comment by MaD70** [comment]
Unfortunately on Windows [Python Standalone Builds requires MS Visual Studio (2017 or later) to build](https://python-build-standalone.readthedocs.io/en/latest/building.html#windows). I have no intention to install that pile of crap.

There is [Static-Python](https://github.com/bendmorris/static-python) which is _a **fork**(!) of CPython that supports building a static interpreter and true standalone executables_. WTF!? It means that the standard interpreter is designed in such a way to actively forbid it. This hostility to static linking when embedding it is outrageously dumb! I have seen this attitude with SDL2 but at least there one has only to set the value of a single macro to statically link it.

I find all this really wasteful and annoying.

**Comment by MaD70** [comment]
I can confirm that [CPython CMake Buildsystem](https://github.com/python-cmake-buildsystem/python-cmake-buildsystem) (CCB) compiles Python 3.6.7 (default) and the executable seems ok.
Some years ago I installed [Microsoft C++ Build Tools](https://visualstudio.microsoft.com/visual-cpp-build-tools/) 2015 to try to compile a project, so I used it now (version 2015 used to be smaller and faster to install than a full Visual Studio).
I created a directory `C:\Python` and executed (**in an MSYS2 shell**, not from Windows command prompt) the following modified instructions from CCB github:
```sh
# Windows
cd /c/Python
mkdir python-build
mkdir python-install
cd python-build
cmake -G "Visual Studio 14 2015" -A x64 BUILD_LIBPYTHON_SHARED=OFF BUILD_EXTENSIONS_AS_BUILTIN=ON \
      CMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX:PATH=/c/Python/python-install ../python-cmake-buildsystem
cmake --build . --config Release # -- /m gave me an error
cmake --build . --config Release --target INSTALL
```

To check that the executable generated is not linking against any dynamic library, apart from system .dll (at least at the time of loading, see [Run-Time Dynamic Linking](https://docs.microsoft.com/en-us/windows/win32/dlls/run-time-dynamic-linking)), I used `ldd`:
```sh
cd /c/Python/python-install/bin
ldd python
      ntdll.dll => /c/WINDOWS/SYSTEM32/ntdll.dll (0x7ff851ff0000)
      KERNEL32.DLL => /c/WINDOWS/System32/KERNEL32.DLL (0x7ff851a40000)
      KERNELBASE.dll => /c/WINDOWS/System32/KERNELBASE.dll (0x7ff84f940000)
      WS2_32.dll => /c/WINDOWS/System32/WS2_32.dll (0x7ff851360000)
      RPCRT4.dll => /c/WINDOWS/System32/RPCRT4.dll (0x7ff850060000)
      SHLWAPI.dll => /c/WINDOWS/System32/SHLWAPI.dll (0x7ff850620000)
      msvcrt.dll => /c/WINDOWS/System32/msvcrt.dll (0x7ff850340000)
      ADVAPI32.dll => /c/WINDOWS/System32/ADVAPI32.dll (0x7ff851990000)
      sechost.dll => /c/WINDOWS/System32/sechost.dll (0x7ff851c90000)
      ucrtbase.dll => /c/WINDOWS/System32/ucrtbase.dll (0x7ff84f740000)
      VERSION.dll => /c/WINDOWS/SYSTEM32/VERSION.dll (0x7ff849820000)
      VCRUNTIME140.dll => /c/WINDOWS/SYSTEM32/VCRUNTIME140.dll (0x7ff83ef50000)
```
The problem is MS compiler usually optimize code less aggressively vs GG/Clang and I'm also not familiar with it, I'm not sure I can proceed with all the experiments I have in mind to test my hypothesis above (compile-time computation + constant propagation are capable enough that they could "peel off" part of the interpreter when the Python program is passed as a constant string **or** as a constant buffer of bytecodes; i.e. that they are approaching [partial evaluation](https://en.wikipedia.org/wiki/Partial_evaluation)).

**Comment by MaD70** [comment]
I can confirm that [CPython CMake Buildsystem](https://github.com/python-cmake-buildsystem/python-cmake-buildsystem) (CCB) compiles Python 3.6.7 (default) and the executable seems ok.
Some years ago I installed [Microsoft C++ Build Tools](https://visualstudio.microsoft.com/visual-cpp-build-tools/) 2015 to try to compile a project, so I used it now (version 2015 used to be smaller and faster to install than a full Visual Studio).
I created a directory `C:\Python` and executed (**in an MSYS2 shell**, not from Windows command prompt) the following modified instructions from CCB github:
```sh
# Windows
cd /c/Python
mkdir python-build
mkdir python-install
cd python-build
cmake -G "Visual Studio 14 2015" -A x64 BUILD_LIBPYTHON_SHARED=OFF BUILD_EXTENSIONS_AS_BUILTIN=ON \
      CMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX:PATH=/c/Python/python-install ../python-cmake-buildsystem
cmake --build . --config Release # -- /m gave me an error
cmake --build . --config Release --target INSTALL
```

To check that the executable generated is not linking against any dynamic library, apart from system .dll (at least at the time of loading, see [Run-Time Dynamic Linking](https://docs.microsoft.com/en-us/windows/win32/dlls/run-time-dynamic-linking)), I used `ldd`:
```sh
cd /c/Python/python-install/bin
ldd python
      ntdll.dll => /c/WINDOWS/SYSTEM32/ntdll.dll (0x7ff851ff0000)
      KERNEL32.DLL => /c/WINDOWS/System32/KERNEL32.DLL (0x7ff851a40000)
      KERNELBASE.dll => /c/WINDOWS/System32/KERNELBASE.dll (0x7ff84f940000)
      WS2_32.dll => /c/WINDOWS/System32/WS2_32.dll (0x7ff851360000)
      RPCRT4.dll => /c/WINDOWS/System32/RPCRT4.dll (0x7ff850060000)
      SHLWAPI.dll => /c/WINDOWS/System32/SHLWAPI.dll (0x7ff850620000)
      msvcrt.dll => /c/WINDOWS/System32/msvcrt.dll (0x7ff850340000)
      ADVAPI32.dll => /c/WINDOWS/System32/ADVAPI32.dll (0x7ff851990000)
      sechost.dll => /c/WINDOWS/System32/sechost.dll (0x7ff851c90000)
      ucrtbase.dll => /c/WINDOWS/System32/ucrtbase.dll (0x7ff84f740000)
      VERSION.dll => /c/WINDOWS/SYSTEM32/VERSION.dll (0x7ff849820000)
      VCRUNTIME140.dll => /c/WINDOWS/SYSTEM32/VCRUNTIME140.dll (0x7ff83ef50000)
```
The problem is MS compiler usually optimize code less aggressively vs GCC/Clang and I'm also not familiar with it, I'm not sure I can proceed with all the experiments I have in mind to test my hypothesis above (compile-time computation + constant propagation are capable enough that they could "peel off" part of the interpreter when the Python program is passed as a constant string **or** as a constant buffer of bytecodes; i.e. that they are approaching [partial evaluation](https://en.wikipedia.org/wiki/Partial_evaluation)).

**Comment by kayhayen** [comment]
@MaD70 mind you, Nuitka wants to embedd the Python interpreter into a binary, and that is another thing alltogether. It's just too complicated to deal with the limitations, esp. as my user base is going to expect that PySide2 and others which deeply tie to CPython specifics just work. I gave up the idea of forking CPython a long time ago as well.

With --onefile solution for Windows, I expect this to be only relevant for performance. But there we will gradually get rid of all CPython API usage, that implies slowness. With MSVC now supporting --lto (which MinGW64 does not) I believe there is more hope in there.

**Comment by kayhayen** [comment]
The improvements I made are released as part of 0.6.10 and require careful selection of the gcc to use, but then it should be good.

**Comment by MaD70** [comment]
@kayhayen sorry for this misunderstanding, it's my fault: I came out of the blue and began speculating without any consideration for the fact that we have probably different backgrounds. So what I wrote (in my improbable English) could be quite cryptic if you are not familiar with source-to-source program transformation techniques like partial evaluation (to which I alluded just in the end of my last comment).

What follow is much longer than I hoped, but it does not require a reply from you. So read it if/when you want/have time. I wrote it down here because it contains some information that could be useful to other people.
<hr />

My idea was **exactly** to avoid any modification to the stock Python interpreter (or limit them to something mild, like source code annotations, that could be easily automated with tools like [Coccinelle](https://coccinelle.gitlabpages.inria.fr/website/)), trying to force any sufficiently advanced C compiler to **approach** a partial evaluator. In particular I was aiming at [the first Futamura projection](https://en.wikipedia.org/wiki/Partial_evaluation#Futamura_projections):

> A particularly interesting example of the use of partial evaluation, first described in the 1970s by Yoshihiko Futamura, is when _prog_ is an interpreter for a programming language.
> 
> If _I_<sub>static</sub> is source code designed to run inside that interpreter, then partial evaluation of the interpreter with respect to this data/program produces _prog*_, a version of the interpreter that only runs that source code, is written in the implementation language of the interpreter, does not require the source code to be resupplied, and runs faster than the original combination of the interpreter and the source. In this case _prog*_ is effectively a compiled version of _I_<sub>static</sub>.
>
> This technique is known as the first Futamura projection…

Of course, a C compiler doesn't emit the transformed source code directly, as a partial evaluator do. It instead specializes a program (with respect to constants known at compile-time) and then proceed to futher transform that to machine code.

Well, in the meantime I ran some simple experiments (eg.: a trivial recursive descent parser for simple arithmetic expressions to which I pass, as a constant known at compile-time, a string with an expression to parse). As much as I tried (but this could be a limit in my knowledge), no matter how I "massage" the original source code, I always clash against some fundamental limitations of current C compilers.

For an example of such limitations see [this simple example of a (recursive) factorial function](https://godbolt.org/z/e5xY9W) at Compiler Explorer: on the left one can see that `fac(18)` is calculated at compiled time and the result is embedded as a constant in the program represented as assembly listing (center-left frame, the instruction `movabs rsi, 6402373705728000`); on the right `fac(19)` triggers a threshold in the compiler (GCC in this case, but Clang is more or less the same) and the factorial is calculated at run-time (center-right frame). Note that even if the function is **not** tail-recursive (the last operation for the `if` branch where x≥2, `x*fac(x-1)`, is the multiplication not the recursive call) the compiler transformed the recursive call into a loop (see the instructions `cmp rax, 1` and `jne .L2`).

So my conclusion is: the more advanced C compilers do some sophisticated program transformations while compiling but these are apparently hard to guide with just annotations and compilers flags (perhaps a compiler plug-in could do wonders). C++ compilers are a different matter, as I already wrote: [`constexpr` and template metaprogramming](https://en.wikipedia.org/wiki/Compile_time_function_execution) are really powerful but they require modification of the source program; partial evaluation instead is (mostly) automatic.

I was thinking about how to further proceed and I have identified two options; both require substantial time and effort (to be clear: **my** time & effort) but the result could be useful in a large number of other projects:
1. resurrect the old, unmaintained [C-Mix, a partial evaluator for C](https://www.itu.dk/people/sestoft/pebook/) (link to the web page of the now freely available book _Partial Evaluation and Automatic Program Generation_; [the sources of C-Mix are in a zip file](https://www.itu.dk/people/sestoft/pebook/Cmix.zip) linked from the section _Online materials related to the book_);
2. improve [LLPE, a partial evaluator for LLVM bitcode](http://www.llpe.org/build/), toward being more cross platform (quoting from [Build Instructions](http://www.llpe.org/build/): "_Building has only been tested under Linux and using the x86-64 architecture; other operating systems and architectures may work, but specialisation may be incorrect due to assumptions about endian-ness and the kernel API used by subject programs._")

Perhaps a third option could be to annotate automatically the C source of the Python stock interpreter with a tool like Coccinelle. Transforming it into a C++ program with `constexpr` and templates in strategic points in such a way that passing it with a Python program (possibly as a buffer of bytecodes resulting from a preliminary pass of the interpreter) as a constant to a C++ compiler has similar effects to partial evaluation and as end result a compilation of the Python program.

I'm still undecided because I don't know how much time I will be able to devote to such a demanding project in the future.
